[mlir][memref]: Fix Bug in GlobalOp Verifier

FranklandJack opened the pull request:
When reconstructing the corresponding tensor type of a memref ensure we include the memory space of the tensor if it exists.

FranklandJack committed:
[mlir][memref]: Fix Bug in GlobalOp Verifier

When comparing the type of the initializer in a `memref::GlobalOp`
against its result only consider the element type and the shape. Other
attributes such as memory space should be ignored since comparing these
between tensors and memrefs doesn't make sense and constructing a memref
in a specific memory space with a tensor that has no such attribute
should be valid.

Signed-off-by: Jack Frankland <jack.frankland@arm.com>

matthias-springer reviewed (changes requested):
This PR assumes that tensor encoding and memref memory space are the same thing. That's not generally the case. Some compilers use that convention during bufferization, but not all do. E.g., the sparse compiler uses the tensor encoding for different purposes.

I don't quite follow what bug this PR is fixing. Why is it desirable that the tensor attribute encoding and the memref memory space match?

FranklandJack commented:
> This PR assumes that tensor encoding and memref memory space are the same thing. That's not generally the case. Some compilers use that convention during bufferization, but not all do. E.g., the sparse compiler uses the tensor encoding for different purposes.
> 
> I don't quite follow what bug this PR is fixing. Why is it desirable that the tensor attribute encoding and the memref memory space match?

Yeah sorry this issue here is a little bit subtle. To take a concerte example, we have an `arith::ConstantOp` which we want to attach an encoding to. Because `arith::ConstantOp` inherits the `AllTypesMatch<["value", "result"]` interface (or trait?) it needs the initializer type to match the value type of the returned tensor, so both need the encoding. During bufferization we pass the `-use-encoding-for-memory-space` and the `arith::ConstantOp` gets lowered to a `memref::GlobalOp` where the initializer is a tensor type with the encoding value and the operations results type is a memref with the corresponding memory space (these are both integers and happen to match when bufferizing with `-use-encoding-for-memory-space`).

The problem is that [`memref::GlobalOp::verify`](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp#L1567-L1577) currently requires the types match but doesn't include the memory space :
```
// Check that the type of the initial value is compatible with the type of
    // the global variable.
    if (auto elementsAttr = llvm::dyn_cast<ElementsAttr>(initValue)) {
      Type initType = elementsAttr.getType();
      Type tensorType = getTensorTypeFromMemRefType(memrefType);
      if (initType != tensorType)
        return emitOpError("initial value expected to be of type ")
               << tensorType << ", but was of type " << initType;
    }
  }
 ```
 So `initType` here includes an encoding and `tensorType` doesn't (even though the `memrefType` it mapped from does include an address space which happens to match the encoding in the case of `-use-encoding-for-memory-space` ).
 
 As an example:
 ```
 func.func @test_constant() -> () {
  %0 = "arith.constant"(){value = dense<[42, 24, 13, 52]> : tensor<4xi32, 1 : i32>} : () -> tensor<4xi32, 1 : i32>
  return
}
```
run with:
```
./build-debug/bin/mlir-opt --one-shot-bufferize="use-encoding-for-memory-space" example.mlir
```
with cause a verification error due to the situation described above.

I feel like there might be other solutions here:
1. Relax the `memref::GlobalOp` verifier to ignore the encoding when it compares the reconstructed tensor type against the initializer.
2. Relax the constraint that the types match exactly on `arith::ContantOp` and allow different encodings for the initializer and the returned value.

FranklandJack commented:
> I would change the `memref.global` verifier: it should check the shape + element type of the memref / tensor types match, nothing more. The only reason why there is a tensor type at all is because we want to reuse `DenseElementsAttr`, which requires a type.

Sounds good to me. I've updated the verifier to do exactly that and added specific tests case for when the element types mismatch and when the shapes mismatch. I've also added a round trip to check we can now construct a memref in a given address space from an initializer with no matching encoding.

Let me know if you would like any further changes.

banach-space commented on mlir/test/Dialect/MemRef/invalid.mlir:
```diff
@@ -342,6 +342,16 @@ memref.global "priate" constant @memref5 : memref<2xf32>  = uninitialized
 // -----
 
+// expected-error @+1 {{op initial value element expected to be of type 'f16', but was of type 'f32'}}
+"memref.global"() <{constant, initial_value = dense<1.000000e+00> : tensor<1xf32>, sym_name = "memref6", sym_visibility = "private", type = memref<1xf16>}> : () -> ()
+
```
Wouldn't a "pretty" version like this work as well?
```suggestion
memref.global "private" constant @memref6 : memref<1xf16> =  dense<[1]>
```

FranklandJack replied:
So I tried this and it gives a slightly different error message which makes me think that in the pretty form it is triggering a different verifier:
```
unexpected decimal integer literal for a floating point value
memref.global "private" constant @memref6 : memref<1xf16>  = dense<[1]>
```

banach-space replied:
OK, thanks for checking. Clearly the attribute is initialised differently in both cases. Since your code only affects the verbose format, the test coverage seems both sufficient and appropriate.

banach-space reviewed (approved):
Thanks, some suggestions inline.

banach-space commented:
@FranklandJack , I noticed that you forced-pushed and erased the history. Please try to avoid that:
* https://llvm.org/docs/GitHub.html#rebasing-pull-requests-and-force-pushes

In particular, some earlier comments refer to the original revision, but that context is gone making it hard for folks referring to this in the future to follow the discussion. 

Not a big deal, it is hard to keep track of all the LLVM policies/guidelines. And, TBH, many people do force-push (e.g. to rebase on top of main), but even then we should preserve the original commits.

Most importantly, thanks for fixing this!

FranklandJack commented on mlir/test/Dialect/MemRef/invalid.mlir:
```diff
@@ -342,6 +342,16 @@ memref.global "priate" constant @memref5 : memref<2xf32>  = uninitialized
+
+// -----
+
+// expected-error @+1 {{op initial value shape expected to be 1, 2 but was 2, 2}}
+"memref.global"() <{constant, initial_value = dense<1.000000e+00> : tensor<2x2xf16>, sym_name = "memref7", sym_visibility = "private", type = memref<1x2xf16>}> : () -> ()
```
I think this might also be exercising a different verifier since the error message being produced is slightly different to the one I'm triggering in my test. Could it be that in the pretty version the type of the initializer is deduced from the return type of the operation vs the type of the dense element attribute?
